### DISCUSSION TITLE
Does "remote_addrs = %any" in IKE connection configurations support IPv6 address?

### DISCUSSION THREAD

**Discussion by yizhao1** [strongswan/strongswan#1072] [General]
Hi,

This is a scenario in which IPv6 hosts use IPsec in transport mode. In my case, there is only one traffic selector in IPv6. When I set "local_addrs = %any" and "remote_addrs = %any" in IKE connections, it doesn't work between IPv6. If I set it to %any6, then the IPv6 can work. My question is does %any support bothe IPv4 and IPv6 or only IPv4 ?  Thanks.
Here are my config files:

swanctl.conf on host:

> connections {
   conn {
      version = 1
      local_addrs = %any6
      remote_addrs = %any6
      proposals = 3des-sha1-modp1024
      rekey_time = 28800s
      local {
         auth = psk
      }
      remote {
         auth = psk
      }
      children {
         child_1 {
            remote_ts = 2001:db8:0:f101::2
            esp_proposals = 3des-sha1
            rekey_time = 2700s
            life_time = 3600s
            mode = transport
            start_action = trap
         }
         icmpv6_rs {
            local_ts = ::/0[ipv6-icmp/133]
            remote_ts = ::/0[ipv6-icmp/133]
            mode = pass
            start_action = trap
         }
         icmpv6_ra {
            local_ts = ::/0[ipv6-icmp/134]
            remote_ts = ::/0[ipv6-icmp/134]
            mode = pass
            start_action = trap
         }
         icmpv6_ns {
            local_ts = ::/0[ipv6-icmp/135]
            remote_ts = ::/0[ipv6-icmp/135]
            mode = pass
            start_action = trap
         }
         icmpv6_na {
            local_ts = ::/0[ipv6-icmp/136]
            remote_ts = ::/0[ipv6-icmp/136]
            mode = pass
            start_action = trap
         }
         icmpv6_r  {
            local_ts = ::/0[ipv6-icmp/137]
            remote_ts = ::/0[ipv6-icmp/137]
            mode = pass
            start_action = trap
         }
      }
   }
}
secrets {
      ike {
         secret = "12345678"
      }
}

swanctl.conf on remote:

> connections {
   conn {
      version = 1
      local_addrs = %any6
      remote_addrs = %any6
      proposals = 3des-sha1-modp1024
      rekey_time = 28800s
      local {
         auth = psk
      }
      remote {
         auth = psk
      }
      children {
         child_1 {
            remote_ts = 2001:db8:0:f101::1
            esp_proposals = 3des-sha1
            rekey_time = 2700s
            life_time = 3600s
            mode = transport
            start_action = trap
         }
         icmpv6_rs {
            local_ts = ::/0[ipv6-icmp/133]
            remote_ts = ::/0[ipv6-icmp/133]
            mode = pass
            start_action = trap
         }
         icmpv6_ra {
            local_ts = ::/0[ipv6-icmp/134]
            remote_ts = ::/0[ipv6-icmp/134]
            mode = pass
            start_action = trap
         }
         icmpv6_ns {
            local_ts = ::/0[ipv6-icmp/135]
            remote_ts = ::/0[ipv6-icmp/135]
            mode = pass
            start_action = trap
         }
         icmpv6_na {
            local_ts = ::/0[ipv6-icmp/136]
            remote_ts = ::/0[ipv6-icmp/136]
            mode = pass
            start_action = trap
         }
         icmpv6_r  {
            local_ts = ::/0[ipv6-icmp/137]
            remote_ts = ::/0[ipv6-icmp/137]
            mode = pass
            start_action = trap
         }
      }
   }
}
secrets {
      ike {
         secret = "12345678"
      }
}

**Comment by tobiasbrunner** [comment]
> My question is does %any support bothe IPv4 and IPv6 or only IPv4 ?

Generally, yes. But what you configured (`remote_addrs` unspecified combined with `mode=transport` and `start_action=trap`) triggers a special scenario. Here `%any` resolves to `%any4` initially (we need a specific address, even if unspecified, i.e. `0.0.0.0` or `::`, to install the trap policy). Later the IKE_SAs are negotiated between the actual addresses from acquires sent by the kernel (i.e. they are determined by the traffic selectors).

The initial address family determines how `local|remote_ts` are resolved if they are not specified. Since your `child_1` sections don't specify `local_ts` it defaults to `dynamic` and that gets resolved to `0.0.0.0/0` if `local_addrs` is `%any` and to `::/0` if it's `%any6`. The former, in turn, prevents trap policies from getting installed because the families of local and remote traffic selectors don't match.

So what you could do is configure `local_ts=::/0`. Or use a specific address if it's known, however, in that case you could also just configure `local_addrs` and `remote_addrs` to those specific addresses and leave `local|remote_ts` unspecified (the special scenario described above is intended for cases where e.g. `remote_ts` is one or more subnets to create tunnels to any of those hosts without having to configure individual connections, not for simple connections between two hosts).

Also note that it's recommended to install bypass policies in a separate connection entry with `local|remote_addrs` set to `127.0.0.1`.

**Reply by yizhao1** [reply]
Sorry for the late response. I added local_ts=::/0 in child_1 and it still did not work.
Here are some logs:

When I set "local_addrs = %any" and "remote_addrs = %any" in IKE connections, there are some errors with child_1 policy installation:

installing 'child_1'
configured proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
adding policy 2001:db8:0:f101::2/128 === ::/0 in [priority 334464, refcount 1]
received netlink error: Invalid argument (22)
unable to add policy 2001:db8:0:f101::2/128 === ::/0 in
adding policy ::/0 === 2001:db8:0:f101::2/128 out [priority 334464, refcount 1]
received netlink error: Invalid argument (22)
unable to add policy ::/0 === 2001:db8:0:f101::2/128 out 
installing trap failed
CHILD_SA child_1{1} state change: CREATED => DESTROYING
deleting policy ::/0 === 2001:db8:0:f101::2/128 out 
received netlink error: No such file or directory (2) 
unable to delete policy ::/0 === 2001:db8:0:f101::2/128 out 
deleting policy 2001:db8:0:f101::2/128 === ::/0 in
received netlink error: No such file or directory (2) 
unable to delete policy 2001:db8:0:f101::2/128 === ::/0 in
installing 'icmpv6_rs'
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] fwd [priority 199679, refcount 1]
installing 'icmpv6_ra'
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] fwd [priority 199679, refcount 1]
installing 'icmpv6_ns'
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] fwd [priority 199679, refcount 1]
installing 'icmpv6_na'
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] fwd [priority 199679, refcount 1]
installing 'icmpv6_r'
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] fwd [priority 199679, refcount 1]


When I set "local_addrs = %any6" and "remote_addrs = %any6" in IKE connections, the child_1 policy can be installed without problems:

installing 'child_1'
configured proposals: ESP:3DES_CBC/HMAC_SHA1_96/NO_EXT_SEQ
adding policy 2001:db8:0:f101::2/128 === ::/0 in [priority 334464, refcount 1]
adding policy ::/0 === 2001:db8:0:f101::2/128 out [priority 334464, refcount 1]
CHILD_SA child_1{1} state change: CREATED => ROUTED
installing 'icmpv6_rs'
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/133] === ::/0[ipv6-icmp/133] fwd [priority 199679, refcount 1]
installing 'icmpv6_ra'
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/134] === ::/0[ipv6-icmp/134] fwd [priority 199679, refcount 1]
installing 'icmpv6_ns'
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/135] === ::/0[ipv6-icmp/135] fwd [priority 199679, refcount 1]
installing 'icmpv6_na'
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/136] === ::/0[ipv6-icmp/136] fwd [priority 199679, refcount 1]
installing 'icmpv6_r'
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] out [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] in [priority 199679, refcount 1]
adding policy ::/0[ipv6-icmp/137] === ::/0[ipv6-icmp/137] fwd [priority 199679, refcount 1]

**Reply by tobiasbrunner** [reply]
It's possible that the kernel enforces that the address family of the template is the same as that of the selector for transport mode. In which case you can't avoid using `%any6`.

**Reply by yizhao1** [reply]
So is there any way to set a value/wildcard for local_addrs/remote_addrs to support both ipv4 and ipv6 in my configuration?

**Reply by tobiasbrunner** [reply]
No, I guess not. But you are also using a special configuration (wildcard traps), and haven't described yet why you are doing so for simple host-to-host connections.

**Comment by supperwangli** [comment]
hi @tobiasbrunner ,

please check the attached patch if is it taken as a fixed patch?(
[0001-strongswan-set-the-family-of-tmpl-to-AF_UNSPEC.patch](https://github.com/strongswan/strongswan/files/13807248/0001-strongswan-set-the-family-of-tmpl-to-AF_UNSPEC.patch)
)

Thanks,
LiWang.
